lib/connection.js

```diff
@@ -168,15 +163,33 @@ Connection.prototype.receiveResponse = function(data) {
 /**
  * Close this connection
  */
-Connection.prototype.close = function() {
+Connection.prototype.close = function(error) {
 
-    if (this.stream) {
+    if (!this.connected) {
+        return;
+    }
+
+    if (this.stream && this.stream.writable) {
         this.stream.end();
     }
+
     this.connected = false;
-    this.disconnected = true;
     this.stream = null;
     this.emit('close');
+
+    var hasRequests = this.requestBuffer.length;
+    if (hasRequests || error) {
+        this.emit('error.server', error || 'Server unexpectedly closed network connection.');
```

Comment:
I worry that it's generally unsafe to retry in this case, but if it were deemed necessary then perhaps a special `error.request` event that receives the original request so that it can be inspected and retried under the right conditions:

```    var hasRequests = this.requestBuffer.length;
    if (hasRequests || error) {
        this.emit('error.server', error || 'Server unexpectedly closed network connection.');
        var request;
        while (--hasRequests >= 0) {
            request = this.requestBuffer.shift();
            this.emit('error.request', request);
        }
    }```